src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs

```diff
@@ -1834,5 +1834,89 @@ void M4((string? s1, string? s2)? n)
                 //             _         => n.Value.ToString(), // 7: n
                 Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "n").WithLocation(31, 26));
         }
+
+        [Fact]
+        [WorkItem(34233, "https://github.com/dotnet/roslyn/issues/34233")]
+        public void SwitchExpressionResultType_01()
+        {
+            CSharpCompilation c = CreateNullableCompilation(@"
+class Test
+{
+    void Test1(int i, object? x, object y)
+    {
+        _ = i switch { 1 => x, _ => y }/*T:object?*/;
```

Comment:
Would it be worth it to add a test with a when clause affecting the state of `x`? For example, `_ = i switch { 1 when x != null => x, _ => y}/*T:object!*/;`